File: ClaseNayelli/C++Nayelli/Stack,Queue,Dequeue/stack.cpp
#include <iostream>
using namespace std;

template<class T>
class Stack{
public:
    
    class Node{
    public:
        Node *next;
        T data;

        Node(){ this->next = nullptr; }

        Node(T data){
            next = nullptr;
            this->data = data;
        }

        ~Node(){
            cout<<"  ~Node("<<data<<")"<<endl;
        }

        friend ostream& operator<<(ostream& output,const Node& n){
            output<< n.data ;
            return output;
        }

    };

    Node *first;

    Stack(){ first = nullptr; }

    ~Stack(){  
        Node* n = this->first;
        while(n != nullptr){
            Node* aux = n;
            n = n->next;
            delete aux;
        }
        cout<<"~Stack Delete\n";
    }

    void push(T data){
        Node* n = new Node(data);
        n->next = this->first;
        this->first = n;
        return;
    }

    T pop(){
        if(this->first == nullptr){
            return (T)NULL;
        }
        Node* aux = this->first;
        this->first = this->first->next;
        T cosa = aux->data;
        delete aux;
        return cosa;
    }

    T top(){
        return this->first == nullptr ? (T)NULL : this->first->data;
    }

    bool isEmpty(){
        return this->first == nullptr ? true : false;
    }

    friend ostream& operator<<(ostream& output, Stack &p){
        output<<"[";
        if(!p.isEmpty()){
            Node* n = p.first;
            while(n != nullptr){
                if(n!=p.first)output<<", ";
                output<<n->data;
                n = n->next;
            }
        }
        output<<"]";
        return output;
    }

};


int main(){
    
    Stack<int> p;
    cout<<p<<'\n';
    for(int i=0;i<10;i++){
        p.push(i);
    }
    cout<<p<<'\n';
    while(!p.isEmpty()){
        cout<<p.top()<<' '<<p<<'\n';
        p.pop();
    }
    cout<<p<<'\n';

    
cout<<"Pruebas\n";
    for(int i=0;i<10;i++) p.push(i);
    cout<<p<<'\n';
    for(int i=0;i<15;i++) cout<<p.pop()<<' '; cout<<'\n';
    cout<<p<<'\n';
    for(int i=0;i<15;i++) cout<<p.top()<<' '; cout<<'\n';
    cout<<p<<'\n';

}

File: EstructurasDeDatosII/PythonNayelli/Stack,Queue,Dequeue/stack.py

class Node:

    def __init__(self):
        self.next = None
        self.data = 0

    def __init__(self,data):
        self.next = None
        self.data = data

    def __str__(self):
        return "Data: "+str(data)


class Stack:

    def __init__(self):
        self.first = None

    def __str__(self):
        s = "Stack: ["
        n = self.first
        while n != None:
            s+=str(n.data)+","
            n = n.next
        s+="]"
        return s;

    def pop(self):
        if self.first == None:
            return False
        aux = self.first 
        self.first = self.first.next
        cosa = aux.data
        del aux
        return cosa

    def push(self,data):
        n = Node(data)
        n.next = self.first
        self.first = n

    def isEmpty(self):
        return True if self.first == None else False

    def top(self):
        return None if self.first == None else self.first.data
    

